src/groth16/prover.rs

```diff
@@ -476,3 +493,294 @@ where
+            &mut multiexp_kern,
+        );
+        let b_g1_aux = multiexp(
+            &worker,
+            b_g1_aux_source,
+            b_aux_density.clone(),
+            aux_assignment.clone(),
+            &mut multiexp_kern,
+        );
+
+        let (b_g2_inputs_source, b_g2_aux_source) = params
+            .lock()
+            .unwrap()
+            .get_b_g2(b_input_density_total, b_aux_density_total)?;
+
+        let b_g2_inputs = multiexp(
+            &worker,
+            b_g2_inputs_source,
+            b_input_density,
+            input_assignment.clone(),
+            &mut multiexp_kern,
+        );
+        let b_g2_aux = multiexp(
+            &worker,
+            b_g2_aux_source,
+            b_aux_density,
+            aux_assignment.clone(),
+            &mut multiexp_kern,
+        );
+
+        (
+            a_inputs,
+            a_aux,
+            b_g1_inputs,
+            b_g1_aux,
+            b_g2_inputs,
+            b_g2_aux,
+        )
+    };
+    debug!("input end");
+
+    debug!("proof start");
+    let proof = {
+        if vk.delta_g1.is_zero() || vk.delta_g2.is_zero() {
+            // If this element is zero, someone is trying to perform a
+            // subversion-CRS attack.
+            return Err(SynthesisError::UnexpectedIdentity);
+        }
+
+        let mut g_a = vk.delta_g1.mul(r);
+        g_a.add_assign_mixed(&vk.alpha_g1);
+        let mut g_b = vk.delta_g2.mul(s);
+        g_b.add_assign_mixed(&vk.beta_g2);
+        let mut g_c;
+        {
+            let mut rs = r;
+            rs.mul_assign(&s);
+
+            g_c = vk.delta_g1.mul(rs);
+            g_c.add_assign(&vk.alpha_g1.mul(s));
+            g_c.add_assign(&vk.beta_g1.mul(r));
+        }
+        let mut a_answer = a_inputs.wait()?;
+        a_answer.add_assign(&a_aux.wait()?);
+        g_a.add_assign(&a_answer);
+        a_answer.mul_assign(s);
+        g_c.add_assign(&a_answer);
+
+        let mut b1_answer = b_g1_inputs.wait()?;
+        b1_answer.add_assign(&b_g1_aux.wait()?);
+        let mut b2_answer = b_g2_inputs.wait()?;
+        b2_answer.add_assign(&b_g2_aux.wait()?);
+
+        g_b.add_assign(&b2_answer);
+        b1_answer.mul_assign(r);
+        g_c.add_assign(&b1_answer);
+        g_c.add_assign(&h.wait()?);
+        g_c.add_assign(&l.wait()?);
+
+        Proof {
+            a: g_a.into_affine(),
+            b: g_b.into_affine(),
+            c: g_c.into_affine(),
+        }
+    };
+    debug!("proof end");
+
+    Ok(proof)
+}
+
+fn create_proof_batch_priority_parallel<E, C, P: ParameterSource<E>>(
+    circuits: Vec<C>,
+    params: P,
+    r_s: Vec<E::Fr>,
+    s_s: Vec<E::Fr>,
+    priority: bool,
+    pool_size: usize,
+) -> Result<Vec<Proof<E>>, SynthesisError>
+where
+    E: Engine,
+    C: Circuit<E> + Send,
+{
+    info!(
+        "create proof batch in parallel mode with pool size {}",
+        pool_size
+    );
+
+    let pool = rayon::ThreadPoolBuilder::new()
```

Comment:
Instead of duplicating the existing code, I would prefer to change the current method, to either use the rayon default pool, or the custom one built here

Reply:
this draft just shows some ideas
agree that duplicating code is useful but awful